source/logrepl/cdc.go

```diff
@@ -23,72 +23,100 @@ import (
 	sdk "github.com/conduitio/conduit-connector-sdk"
 	"github.com/jackc/pglogrepl"
 	"github.com/jackc/pgx/v5/pgconn"
-	"github.com/jackc/pgx/v5/pgxpool"
 )
 
 // Config holds configuration values for CDCIterator.
-type Config struct {
-	Position        sdk.Position
+type CDCConfig struct {
+	LSN             pglogrepl.LSN
 	SlotName        string
 	PublicationName string
 	Tables          []string
 	TableKeys       map[string]string
 }
 
-func (c Config) LSN() (pglogrepl.LSN, error) {
-	if len(c.Position) == 0 {
-		return 0, nil
-	}
-	lsn, err := PositionToLSN(c.Position)
-	if err != nil {
-		return 0, fmt.Errorf("failed to parse position: %w", err)
+func (c Config) Validate() error {
+	var errs []error
+	// make sure we have all table keys
+	for _, tableName := range c.Tables {
+		if c.TableKeys[tableName] == "" {
+			errs = append(errs, fmt.Errorf("missing key for table %q", tableName))
+		}
 	}
-	return lsn, nil
+
+	return errors.Join(errs...)
 }
 
 // CDCIterator asynchronously listens for events from the logical replication
 // slot and returns them to the caller through Next.
 type CDCIterator struct {
-	config  Config
+	config  CDCConfig
 	records chan sdk.Record
+	pgconn  *pgconn.PgConn
 
 	sub *internal.Subscription
 }
 
 // NewCDCIterator sets up the subscription to a logical replication slot and
 // starts a goroutine that listens to events. The goroutine will keep running
 // until either the context is canceled or Teardown is called.
-func NewCDCIterator(ctx context.Context, connPool *pgxpool.Pool, config Config) (*CDCIterator, error) {
-	i := &CDCIterator{
-		config:  config,
-		records: make(chan sdk.Record),
+func NewCDCIterator(ctx context.Context, pgconf *pgconn.Config, c CDCConfig) (*CDCIterator, error) {
+	conn, err := pgconn.ConnectConfig(ctx, withReplication(pgconf))
+	if err != nil {
+		return nil, fmt.Errorf("could not establish replication connection: %w", err)
 	}
 
-	err := i.attachSubscription(connPool.Config().ConnConfig.Config)
-	if err != nil {
-		return nil, fmt.Errorf("failed to setup subscription: %w", err)
+	if err := internal.CreatePublication(
```

Comment:
I don't get it, what's the reason behind pushing this logic from `internal.Subscription` into the CDC iterator?

Reply:
I don't think it belongs there, `publication` and `subscriptions` are two different objects. Albeit a sub requiring a pub, shouldn't manage it. Since `subscription` only responsibility is read CDC events. 

The CDC iterator in turn can manage the publication.